Added composite Key feature

leonmezu1 opened the pull request:
# Pull Request for Final Project: Milestone 6

   - Friendship table now has double column primary key (Composite Key) accounting user and rqstuser

leonmezu1 committed:
Added composite Key feature

leonmezu1 committed:
Added new test

leonmezu1 committed:
Raise exception at unvalid frienship creation

Usagib committed:
Two column friendships added to friendship model

Usagib commented:
Added two column relation to Friendship model. re sendig pr for code review

mosaaleb reviewed (changes requested):
#### Hi @leonmezu1, @Usagib 

I got this error when I tried to accept a friendship request

![error](https://user-images.githubusercontent.com/4246637/77802264-2a3eb200-7083-11ea-90ab-cb512a1c12cc.png)

- [x] **[OPTOINAL]** Class methods are usually used for doing bulk queries i.e. (getting all accepted friendships, all posts which have at least 3 comments, ....),  scopes can be used to do the same. But when you operate on a single record it's better to use instance methods.

When the functionality you are implementing is tied to the identity of a particular object, you should be using instance methods.

- [x] There are some methods in the friendship model that are not tested yet -- let's add them.

- [x] `friends` and `pending_friends` method on user model are better to implement using association (active record queries) instead of relying on ruby enumerable because database queries are faster than using enumerable.

This can be achieved by using `has_many through`
```ruby
has_many :friendships, foreign_key: 'user_id', dependent: :delete_all
has_many :friends, -> { where(friendships: { status: true }) }, through: :friendships, source: :rqstuser

# the same can be applied for `pending_friends`
```


Please handle these requirements and submit a new code review.
Feel free to reach me on slack if you have any questions.

##### Muhammad Ebeid

leonmezu1 committed:
Required fixes implementation done

As part of the code tunning, TSE's requirements were fulfilled fixing app functionality, avoiding class methods where possible, replacing deprecated methods with scopes and added, refactored extra features testing

leonmezu1 commented:
Previous requirements are now implemented

samgaco commented on app/models/user.rb:
```diff
@@ -43,18 +28,6 @@ def friend_requests
     inverse_friendships.map { |fr| fr.user unless fr.status }.compact
   end
 
```
>  def friend_requests
>     inverse_friendships.map { |fr| fr.user unless fr.status }.compact
>   end

isn't that the same as pending friends?, remember DRY

>   has_many :pending_friends, -> { where(friendships: { status: nil }) }, through: :friendships, source: :rqstuser

leonmezu1 replied:
After a code inspection, it's been detected that the scope pending_friends lacks the capability of taking the inverse_friendships attribute from user type objects, which is a necessary condition to retrieve the friend requests users list, because of this, it's been decided to remove this scope. 

@Usagib  @leonmezu1  :octocat:

samgaco reviewed (changes requested):
Hi @leonmezu1 and @Usagib :high_brightness: 

The previous requirements are working now, but there is something still missing, the button to accept/reject friendships should also appear in the all users page:

- [ ] There is a button to invite user to friendship on users list page and on single user page ([screenshot](https://gitlab.com/microverse/guides/projects/requirements_screenshots/raw/master/images/rails/final_project/milestone5/user_can_see_invite_buttton.png) from [Pathwright](https://microverse.pathwright.com/library/fast-track-curriculum/69047/path/step/49736080/)) 



----------------

* ###### STATUS OF THE MILESTONE: REQUESTED CHANGES :stop_sign:

leonmezu1 committed:
Required changes done

mcrd25 reviewed (approved):
# changes approved 💪 🎊 

Hey @leonmezu1! Hey @Usagib! 👋 

Good job implementing the change requested from last review! 
Because you've successfully implemented the changes requested by previous reviews there s no need for further changes, you may now merge to development! 😄 

***If you have any questions, now or in the future, do not hesitate to contact me on social media, email or on slack and I will be glad to chat with you.***

*Happy coding 😄,*
**[--Maya](https://github.com/mcrd25)**
[GitHub](https://github.com/mcrd25) [Twitter](https://twitter.com/mcrd25) [linkedin](https://www.linkedin.com/in/mayadouglas/) [GitLab](https://gitlab.com/mcrd25)

leonmezu1 committed:
Final push, redirect back route applied
